skyrl-train/skyrl_train/utils/trainer_utils.py

```diff
@@ -235,6 +236,258 @@ def dump_per_dataset_eval_results(
     print(f"Dumped aggregated eval metrics to {aggregated_filename}")
 
 
+def handle_dynamic_sampling(
+    generator_output: GeneratorOutput,
+    uids: List[str],
+    sampling_config: Dict[str, Any],
+    collected_state: Optional[Dict[str, Any]] = None,
+) -> Tuple[GeneratorOutput, List[str], bool, Optional[Dict[str, Any]]]:
+    """
+    Handle dynamic sampling with different strategies (filter, replace).
+
+    filter (used in DAPO) - filter out groups with std == 0 and group size > 1 and resample until we have enough prompts
+    replace (used in POLARIS, WebSailor) - replace bad (std == 0) samples with good (std > 0) samples
+
+    Args:
+        generator_output: Current batch generator output
+        uids: Current batch UIDs
+        sampling_config: Configuration dict with sampling parameters
+        collected_state: State for accumulating data across batches (for filter strategy)
+
+    Returns:
+        Tuple of (processed_generator_output, processed_uids, keep_sampling, updated_state)
+    """
+    sampling_type = sampling_config.get("type", None)
+
+    if sampling_type is None:
+        return generator_output, uids, False, None
+
+    if sampling_type == "replace":
+        # For "replace" strategy, the collected state is not used.
+        processed_output, processed_uids, keep_sampling = handle_replace_sampling(
+            generator_output, uids, sampling_config
+        )
+        return processed_output, processed_uids, keep_sampling, collected_state
+    elif sampling_type == "filter":
+        # For filter strategies, accumulate the generator output and UIDs across batches in collected_state if we are sampling repeatedly.
+        return handle_filter_sampling(generator_output, uids, sampling_config, collected_state)
+    else:
+        raise ValueError(f"Invalid dynamic sampling type: {sampling_type}")
+
+
+def handle_replace_sampling(
+    generator_output: GeneratorOutput, uids: List[str], sampling_config: Dict[str, Any]
+) -> Tuple[GeneratorOutput, List[str], bool]:
+    """
+    Handle replace sampling strategy based on POLARIS implementation (https://github.com/ChenxinAn-fdu/POLARIS/).
```

Comment:
Can you actually link to the core replace sampling loop itself? It wil be easier to compare for users 

```suggestion
    Handle replace sampling strategy based on POLARIS implementation (https://github.com/ChenxinAn-fdu/POLARIS/blob/8c82adb16b8e45c1a34f6d0e23e35deb66dd1ae7/verl/verl/trainer/ppo/ray_trainer.py#L995-L1022).
```